### PR TITLE
Add terminal-based steering interface via an extract

### PR DESCRIPTION
**Pull request by siramok**
This PR adds a new extract which puts the user into a basic shell-like steering interface. The main benefit is to have a simpler, no-dependency alternative to the existing ascent-jupyter-bridge steering interface.

It can be accessed via:
```-
  action: "add_extracts"
  extracts:
    e1:
      type: "steering"
```
The current implementation allows users to manually, at runtime:
- Add, edit, or delete numeric/string parameters with which to run callbacks (via an internally managed conduit node).
- List registered callbacks (via new APIs that expose the callback names).
- Run callbacks by name.
- See any resulting output from running a callback.

Still to-do:
- [x] Verify that all MPI usage is properly guarded by ASCENT_MPI_ENABLED.
- [x] Similarly, verify that things work properly without MPI.
- [x] Add tests.

Despite not being ready, the big picture implementation details are unlikely to change (where I've implemented it, how I've plumbed it through). Any feedback you may have for me is welcome.

### CODE REVIEW

File: src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+
+    // Descriptions
+    descriptions["exit"] = "Exit the shell.";
+    descriptions["help"] = "Print this help message.";
+    descriptions["list"] = "List all registered Ascent callbacks.";
+    descriptions["param"] = "Prints the current parameters. Can set, modify, or parameters with 'param add|edit|remove key value.";
+    descriptions["run"] = "Run an Ascent callback with 'run my_callback_name'.";
+
+#ifdef ASCENT_MPI_ENABLED
+    MPI_Comm_rank(MPI_COMM_WORLD, &m_rank);
```
Comment (cyrush) [inline src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp]: We might not be using COMM_WORLD, so let's grab the right communicator using:
```
MPI_Comm mpi_comm = MPI_Comm_f2c(flow::Workspace::default_mpi_comm());
```

File: src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.hpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    ~Steering();
+    virtual void declare_interface(conduit::Node &i);
+    virtual bool verify_params(const conduit::Node &params,
+                               conduit::Node &info);
+    virtual void execute();
+private:
+    std::map<std::string, std::function<void()>> commands;
+    std::map<std::string, std::string> descriptions;
+    conduit::Node params;
+    conduit::Node output;
```
Comment (cyrush) [inline src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.hpp]: We like to use `m_` prefix for member vars

File: src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.hpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    int m_rank;
+    bool running;
+    bool empty_params;
+
+    void empty_run();
+    void exit_shell();
+    void parse_input(std::string cmd, std::vector<std::string> args);
```
Comment (cyrush) [inline src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.hpp]:
```suggestion
    void parse_input(std::string &cmd, std::vector<std::string> &args);
```

File: src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.hpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    void list_callbacks();
+    void modify_params(std::vector<std::string> args);
```
Comment (cyrush) [inline src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.hpp]:
```suggestion
    void modify_params(std::vector<std::string> &args);
```

File: src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.hpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    void print_help();
+    void print_params();
+    void run_callback(std::vector<std::string> args);
```
Comment (cyrush) [inline src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.hpp]:
```suggestion
    void run_callback(std::vector<std::string> &args);
```

File: src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+        if (!has_callback)
+        {
+            return;
+        }
+
+        ascent::execute_callback(callback, params, output);
+
+        // This might not be necessary?
+#ifdef ASCENT_MPI_ENABLED
+        MPI_Barrier(MPI_COMM_WORLD);
```
Comment (cyrush) [inline src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp]: fetch comp from workspace

File: src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+            std::cout << std::endl << "None" << std::endl << std::endl;
+        }
+        else
+        {
+            params.print();
+        }
+    }
+}
+
+void Steering::run_callback(std::vector<std::string> args)
```
Comment (cyrush) [inline src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp]:
```suggestion
void Steering::run_callback(std::vector<std::string> &args)
```

File: src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+            for (const auto &callback : bool_callback_names)
+            {
+                std::cout << callback << " ";
+            }
+        }
+        std::cout << std::endl << std::endl;
+    }
+}
+
+void Steering::modify_params(std::vector<std::string> args)
```
Comment (cyrush) [inline src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp]:
```suggestion
void Steering::modify_params(std::vector<std::string> &args)
```

File: src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+        
+        if (m_rank == 0)
+        {
+            std::cout << "steering>" << std::endl;
+            std::getline(std::cin, input);
+            input_size = input.size();
+        }
+
+#ifdef ASCENT_MPI_ENABLED
+        MPI_Bcast(&input_size, 1, MPI_INT, 0, MPI_COMM_WORLD);
```
Comment (cyrush) [inline src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp]: grab comm from workspace

File: src/libs/ascent/ascent.hpp [hunk 10]
```diff
@@ -87,6 +87,14 @@ void        ASCENT_API execute_callback(std::string callback_name,
 //-----------------------------------------------------------------------------
 bool        ASCENT_API execute_callback(std::string callback_name);
 
+//-----------------------------------------------------------------------------
+std::vector<std::string>
```
Comment (cyrush) [inline src/libs/ascent/ascent.hpp]: ref copy out style will be slight more efficient:
```
void ASCENT_API get_void_callbacks(std::vector<std::string> &name);
```

File: src/libs/ascent/ascent.hpp [hunk 11]
```diff
@@ -87,6 +87,14 @@ void        ASCENT_API execute_callback(std::string callback_name,
...
+            ASCENT_API get_void_callbacks();
+
+//-----------------------------------------------------------------------------
+std::vector<std::string>
+            ASCENT_API get_bool_callbacks();
```
Comment (cyrush) [inline src/libs/ascent/ascent.hpp]: ref copy out style will be slight more efficient:
```
void ASCENT_API get_bool_callbacks(std::vector<std::string> &name);
```

File: src/libs/ascent/ascent.hpp [hunk 12]
```diff
@@ -87,6 +87,14 @@ void        ASCENT_API execute_callback(std::string callback_name,
 //-----------------------------------------------------------------------------
 bool        ASCENT_API execute_callback(std::string callback_name);
```
Comment (cyrush) [inline src/libs/ascent/ascent.hpp]:
```suggestion
bool        ASCENT_API execute_callback(const std::string &callback_name);
```

File: src/libs/ascent/ascent.cpp [hunk 13]
```diff
@@ -928,6 +928,28 @@ execute_callback(std::string callback_name)
     return callback_function();
 }
 
+//-----------------------------------------------------------------------------
+std::vector<std::string>
```
Comment (cyrush) [inline src/libs/ascent/ascent.cpp]: copy out still ref args are a bit more efficient (see below)

**Review by cyrush — COMMENTED** [review]
Look great!
a few minor suggestions related to using C++ refs for strings and vectors, as well as the recipe for grabbing the right MPI comm

**Review by cyrush — APPROVED** [review]
thanks for the tweaks!

**Comment by cyrush** [comment]
Tests exercising steering look good -  let me know when you feel its ready to merge.

**Comment by siramok** [comment]
Sweet, thanks Cyrus. I just made another polishing pass. Added comments, fixed formatting, simplified logic where I could, and made some tweaks after testing it out for myself. I feel comfortable merging it in it's current state.

I'll go create a new branch to work on an example instrumentation which makes use of this.

**Pull request merged**